glusterd2/commands/volumes/volume-smartvol-txn.go

```diff
@@ -153,8 +173,148 @@ func txnUndoPrepareBricks(c transaction.TxnCtx) error {
 				}).Error("thinpool remove failed")
 			}
 
-			// Update current Vg free size
-			deviceutils.UpdateDeviceFreeSize(gdctx.MyUUID.String(), b.RootDevice)
+			var freesizeSet bool
+			key := "freesizeSet." + b.PeerID + b.Path
+			if err := c.Get(key, &freesizeSet); err != nil {
+				c.Logger().WithError(err).WithField("key", key).Error("failed to get key from store")
+				return err
+			}
+
+			// Reset Free size only if freeSize is set in transaction
+			if freesizeSet {
+				deviceutils.AddDeviceFreeSize(gdctx.MyUUID.String(), b.RootDevice, b.TotalSize)
+			}
+		}
+	}
+
+	return nil
+}
+
+// PrepareBrickLoop prepares a single brick
+func PrepareBrickLoop(b api.BrickReq, c transaction.TxnCtx) error {
+	if b.PeerID != gdctx.MyUUID.String() {
+		return nil
+	}
+
+	if err := c.Set("freesizeSet."+b.PeerID+b.Path, false); err != nil {
+		return err
+	}
+
+	// Create Mount directory
+	mountRoot := strings.TrimSuffix(b.Path, b.BrickDirSuffix)
+	err := os.MkdirAll(mountRoot, os.ModeDir|os.ModePerm)
+	if err != nil {
+		c.Logger().WithError(err).WithField("path", mountRoot).Error("failed to create brick mount directory")
+		return err
+	}
+
+	// Dir creation
+	err = os.MkdirAll(path.Dir(b.DevicePath), 0700)
+	if err != nil {
+		c.Logger().WithError(err).WithField("device", path.Dir(b.DevicePath)).Error("brick device directory create failed")
+		return err
+	}
+
+	// Loop file creation
+	loopFile, err := os.OpenFile(b.DevicePath, os.O_RDWR|os.O_CREATE, 0700)
+	if err != nil {
+		c.Logger().WithError(err).WithField("device", b.DevicePath).Error("brick device create failed")
+		return err
+	}
+	defer loopFile.Close()
+	err = os.Truncate(b.DevicePath, int64(b.Size))
```

Comment:
I am inclined to setup max size of single brick in this model to be 100GB to start with. Lets do more testing and get to valid number later. But 100GB sounds reasonable to start with.

`Max-brick-size: 100GB`

So if users ask for bigger sized volumes, we need to create a volume with DHT.

If it is block-hosting-volume, then, we need to setup shard-size of 2GB (mainly for reducing overhead of self-heal etc), and also to handle big sized RWO volume. Even there, we can have request for BHV, based on RWO request? because we are creating some 5GB volume now? (@poornimag )

Reply:
To enable sharding for block hosting volume, `Options` needs to be populated here https://github.com/gluster/glusterd2/blob/master/plugins/blockvolume/hostvol_opts.go#L46

```
"Options": {
    "features/shard.shard": "on",
    "features/shard.shard-block-size": "2G",
}
```

@poornimag @oshankkumar ^^